lgn-worker/src/main.rs

```diff
@@ -114,12 +129,195 @@ fn main() -> anyhow::Result<()> {
+            run(&config)?;
+            Ok(())
+        })?;
+    }
+
+    Ok(())
 }
 
-fn run(config: &Config) -> Result<()> {
-    info!("Version: {}", env!("CARGO_PKG_VERSION"));
-    let lagrange_wallet = match (
+async fn maybe_download_checksum(config: &Config) -> Result<()> {
+    if config.public_params.skip_checksum {
+        return Ok(());
+    }
+
+    // Fetch checksum file
+    // The checksum file can be generated in two ways.
+    // 1- Run the worker, and it will download and spit out the checksum on disk
+    // 2- Manually download the params then install with the checksums bin crate and run checksums -c -r zkmr_params -a BLAKE3
+    let checksum_url = &config.public_params.checksum_url;
+    let expected_checksums_file = &config.public_params.checksum_expected_local_path;
+
+    let response = reqwest::get(checksum_url)
+        .await
+        .context("Failed to fetch checksum file")?
+        .text()
+        .await
+        .context("Failed to read response text")?;
+
+    let mut file = tokio::fs::File::create(expected_checksums_file)
+        .await
+        .context("Failed to create local checksum file")?;
+
+    file.write_all(response.as_bytes())
+        .await
+        .context("Failed to write checksum file")?;
+
+    drop(file);
+
+    Ok(())
+}
+
+async fn run_with_grpc(config: &Config, grpc_url: &str) -> Result<()> {
+    let uri = grpc_url.parse::<tonic::transport::Uri>()?;
+    let (mut outbound, outbound_rx) = tokio::sync::mpsc::channel(1024);
+
+    info!("Verifying the checksums");
+
+    maybe_download_checksum(config).await?;
+
+    let mut provers_manager =
+        tokio::task::block_in_place(move || -> Result<ProversManager<TaskType, ReplyType>> {
+            let mut provers_manager = ProversManager::<TaskType, ReplyType>::new();
+            register_v1_provers(config, &mut provers_manager)
+                .context("while registering provers")?;
+            Ok(provers_manager)
+        })?;
+
+    if !config.public_params.skip_checksum {
+        verify_directory_checksums(
+            &config.public_params.dir,
+            &config.public_params.checksum_expected_local_path,
+        )
+        .context("Failed to verify checksums")?;
+    }
+
+    info!("Connecting to GW at uri {uri}");
+    let outbound_rx = tokio_stream::wrappers::ReceiverStream::new(outbound_rx);
+
+    let wallet = get_wallet(config)?;
+    let claims = get_claims(config)?;
+    let token = JWTAuth::new(claims, &wallet)?.encode()?;
+
+    let channel = tonic::transport::Channel::builder(uri).connect().await?;
+    let token: MetadataValue<_> = format!("Bearer {token}").parse()?;
+
+    let message_size = 64 * 1024 * 1024;
+
+    let mut client = lagrange::workers_service_client::WorkersServiceClient::with_interceptor(
+        channel,
+        move |mut req: Request<()>| {
+            req.metadata_mut().insert("authorization", token.clone());
+            Ok(req)
+        },
```

Comment:
Question: Can we make the authentication per request? To prevent replay attacks? Or perhaps is the idea to use this over an https channel?

Reply:
So with bidirectional streaming there will be only one request and the result of that will be two streams of messages. The metadata including the auth can only be fetched from the request. But honestly I do not see it work any other way then over a secure channel.